hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,563 @@
+
+        final var syntheticCryptoTransferTxn =
+                TransactionBody.newBuilder().cryptoTransfer(body).build();
+        final var transferContext = new TransferContextImpl(context, body, true);
+        CryptoTransferExecutor.chargeCustomFee(syntheticCryptoTransferTxn, transferContext);
+    }
+
+    /**
+     *  Create new {@link AccountAirdrop} and if the sender has already existing pending airdrops
+     *  link them together and update the account store and the pending airdrop store with the new values
+     */
+    private AccountAirdrop getNewAccountAirdropAndUpdateStores(
+            Account senderAccount,
+            PendingAirdropId pendingId,
+            PendingAirdropValue pendingValue,
+            WritableAccountStore accountStore,
+            WritableAirdropStore pendingStore) {
+
+        AccountAirdrop newAccountAirdrop;
+        if (senderAccount.hasHeadPendingAirdropId()) {
+            // Get the previous head pending airdrop and update the previous airdrop ID
+            var headAirdropId = senderAccount.headPendingAirdropIdOrThrow();
+            var headAccountAirdrop = pendingStore.getForModify(headAirdropId);
+            validateTrue(headAccountAirdrop != null, INVALID_TOKEN_ID);
+            var updatedAirdrop =
+                    headAccountAirdrop.copyBuilder().previousAirdrop(pendingId).build();
+            pendingStore.patch(headAirdropId, updatedAirdrop);
+
+            // Create new account airdrop with next airdrop ID the previous head airdrop
+            newAccountAirdrop = createAccountAirdrop(pendingId, pendingValue, headAirdropId);
+        } else {
+            newAccountAirdrop = createAccountAirdrop(pendingId, pendingValue);
+        }
+        // Update the sender account with new head pending airdrop
+        var updatedSenderAccount =
+                senderAccount.copyBuilder().headPendingAirdropId(pendingId).build();
+        accountStore.put(updatedSenderAccount);
+        return newAccountAirdrop;
+    }
+
+    /**
+     *  Calculate the fees for the token airdrop transaction. The fees are calculated by combining the
+     *  default airdrop fees, the crypto transfer fees and the token association fees.
+     */
+    @NonNull
+    @Override
+    public Fees calculateFees(@NonNull final FeeContext feeContext) {
+        final var op = feeContext.body().tokenAirdropOrThrow();
+        final var tokensConfig = feeContext.configuration().getConfigData(TokensConfig.class);
+        validateTrue(tokensConfig.airdropsEnabled(), NOT_SUPPORTED);
+
+        final var defaultAirdropFees =
+                feeContext.feeCalculatorFactory().feeCalculator(SubType.DEFAULT).calculate();
+        final var cryptoTransferFees = calculateCryptoTransferFees(feeContext, op.tokenTransfers());
+        final var tokenAssociationFees = calculateTokenAssociationFees(feeContext, op);
+        return combineFees(List.of(defaultAirdropFees, cryptoTransferFees, tokenAssociationFees));
```

Comment:
This should just be doing airdrop fees right ? `Transfer` fee only happens if it successfully is transferred ?

Reply:
That would be easier but unfortunately, we should pre-pay for everything. It does matter if the transfer happens or not: https://github.com/hashgraph/hedera-services/pull/12794/files#diff-abfa443cad3b3b701ac6c08524f6be2af3a76d9b7a62b56c0f3ae330331c4c59R191